File: scripts/lib/app-version.js
// «Har jeg siste versjon?» — the published half of the iOS build-stamp
// contract. The pipeline publishes docs/data/app-version.json carrying the
// short hash of the last commit that touched ios/; the sideloaded app
// compares it against its own build-time Info.plist stamp (see
// ios/Zenji/BuildStamp.swift). Data commits (docs/data/) never touch ios/,
// so the hash only moves when the app's source actually changed.
import { execFileSync } from "child_process";
import fs from "fs";
import path from "path";

// Reads {iosCommit, committedAt} from git, or null when git/history is
// unavailable (e.g. a tarball checkout) — the pipeline then simply skips the
// file and the app keeps showing its stamp without a verdict. `exec` is
// injectable for tests.
export function readIosCommit(repoRoot, exec = execFileSync) {
	try {
		const out = exec("git", ["-C", repoRoot, "log", "-1", "--format=%h %cI", "--", "ios/"], {
			encoding: "utf-8",
		}).trim();
		if (!out) return null;
		const [iosCommit, committedAt] = out.split(" ");
		if (!iosCommit) return null;
		return { iosCommit, committedAt: committedAt || null };
	} catch {
		return null;
	}
}

// Reads scripts/config/testflight.json — the record of the LAST UPLOADED
// TestFlight build (written by scripts/record-testflight.js right after an
// upload; lives OUTSIDE ios/ so recording it never moves the ios/-commit
// stamp it records). Returns null when absent/invalid: app-version.json then
// simply carries no testflight block and the app falls back to the pure
// commit comparison.
export function readTestflight(repoRoot, fsImpl = fs) {
	try {
		const raw = fsImpl.readFileSync(path.join(repoRoot, "scripts", "config", "testflight.json"), "utf-8");
		const parsed = JSON.parse(raw);
		if (!parsed || typeof parsed.stamp !== "string" || !parsed.stamp) return null;
		return parsed;
	} catch {
		return null;
	}
}

// Deliberately NO generatedAt/run-timestamp: the file must be byte-identical
// between runs on unchanged input (manifest idempotence — and the hourly
// pipeline would otherwise commit + deploy churn every run). It changes when
// and only when the last ios/-touching commit changes — or a new TestFlight
// build is recorded.
export function buildAppVersion(commit, testflight = null) {
	if (!commit) return null;
	const out = {
		iosCommit: commit.iosCommit,
		committedAt: commit.committedAt,
	};
	if (testflight) {
		out.testflight = {
			stamp: testflight.stamp,
			build: testflight.build ?? null,
			version: testflight.version ?? null,
			uploadedAt: testflight.uploadedAt ?? null,
		};
	}
	return out;
}

File: ios/Sportivista/Sync/AppVersion.swift
//
//  AppVersion.swift
//  Sportivista
//
//  «Har jeg siste versjon?» — the published half. The static pipeline writes
//  docs/data/app-version.json with the short hash of the last commit that
//  touched ios/, and the ordinary manifest sync delivers it here like any
//  other data file (it is in SyncClient.defaultFilesOfInterest). The app
//  compares it against its own build-time stamp (BuildStamp, app target) —
//  no extra networking, no store, no server.
//
//  Lives in Sync/ (not the app root) because DataStore is compiled into the
//  widget too, and every type DataStore touches must be widget-safe.
//

import Foundation

struct AppVersion: Codable, Equatable, Sendable {
    /// Short hash of the last commit touching ios/ on main.
    var iosCommit: String
    /// ISO timestamp of that commit (informational).
    var committedAt: String?
    /// When the pipeline published this file (informational).
    var generatedAt: String?
    /// The last recorded TestFlight upload (WP-17) — absent until the first
    /// recorded upload, and in dev-only checkouts.
    var testflight: TestFlightVersion? = nil
}

struct TestFlightVersion: Codable, Equatable, Sendable {
    /// The ios/-commit stamp the uploaded archive was built from.
    var stamp: String
    /// TestFlight build number (informational).
    var build: Int? = nil
    /// Marketing version (informational).
    var version: String? = nil
    var uploadedAt: String? = nil
}

/// The pure «har jeg siste versjon?»-judgement — lives here (Sync/, compiled
/// into SportivistaTests) so it is unit-testable; the Bundle-reading glue in
/// BuildStamp (app root) just delegates.
enum AppVersionCheck {
    /// The build's own stamp, written into the built Info.plist by the
    /// post-build script in ios/project.yml. "ukjent" in builds that skipped
    /// the script (previews, hostless test bundles).
    static var bundleStamp: String {
        (Bundle.main.object(forInfoDictionaryKey: "SportivistaBuildStamp") as? String) ?? "ukjent"
    }

    static var bundleDate: String {
        (Bundle.main.object(forInfoDictionaryKey: "SportivistaBuildDate") as? String) ?? ""
    }

    /// The foot line for THIS running build.
    static func footLine(published: AppVersion?) -> String {
        line(stamp: bundleStamp, date: bundleDate, published: published)
    }

    /// `nil` = cannot judge (no published truth yet, or the build is
    /// unstamped). A `-dirty` build compares on its base commit — it is
    /// "current" in the honest sense of "built on top of the latest".
    ///
    /// Current = the newest code OR the newest SHIPPABLE build (WP-17): a
    /// TestFlight install cannot be more current than the last recorded
    /// upload, so ios/-commits that haven't shipped as a build must not nag
    /// testers with «NYERE FINNES».
    static func isCurrent(stamp: String, published: AppVersion?) -> Bool? {
        guard let latest = published?.iosCommit, !latest.isEmpty,
              stamp != "ukjent", !stamp.isEmpty else { return nil }
        let base = stamp.hasSuffix("-dirty") ? String(stamp.dropLast("-dirty".count)) : stamp
        if base == latest { return true }
        if let shipped = published?.testflight?.stamp, !shipped.isEmpty, base == shipped { return true }
        return false
    }

    /// The quiet foot line: «BYGG a1b2c3d · 16.07 21:40 · SISTE»,
    /// «… · NYERE FINNES (d4e5f6a)» or just the stamp when there is no verdict.
    static func line(stamp: String, date: String, published: AppVersion?) -> String {
        let head = date.isEmpty ? "BYGG \(stamp)" : "BYGG \(stamp) · \(date)"
        switch isCurrent(stamp: stamp, published: published) {
        case .some(true): return head + " · SISTE"
        case .some(false): return head + " · NYERE FINNES (\(published?.iosCommit ?? "?"))"
        case .none: return head
        }
    }
}
